Create `fleetctl` pkg asset every release

allenhouchins opened the pull request:
For #40518 
For #42972 

Migrating this functionality from my personal GitHub repo to native Fleet functionality: https://github.com/allenhouchins/latest-fleetctl-package

This pull request introduces an automated process for building, signing, notarizing, and distributing a macOS `.pkg` installer for the `fleetctl` command-line tool. The changes ensure that every release includes a properly signed and notarized installer. The main improvements are the addition of a reusable packaging script, integration with the main release pipeline, and a dedicated workflow for manual testing.

**Key changes include:**

### Automated macOS `.pkg` creation and signing

* Added a new script `tools/build-fleetctl-pkg/main.sh` to automate the process of creating, signing, notarizing, and uploading the macOS `.pkg` installer for `fleetctl`. This script checks for required environment variables, handles retries for notarization, and uploads the package to the GitHub release.

### Release workflow integration

* Updated the main release workflow (`.github/workflows/goreleaser-fleet.yaml`) to:
  - Upload the signed universal `fleetctl` binary as an artifact after building.
  - Add a new `build-fleetctl-pkg` job that downloads the signed binary, verifies it, and invokes the new packaging script to produce and upload the notarized `.pkg` installer as part of the production release process.

* Updated the GoReleaser configuration (`.goreleaser.yml`) to run the new packaging script as a post-signing hook, ensuring the `.pkg` is built automatically on macOS environments after the binary is signed.

### Manual testing workflow

* Introduced a dedicated workflow (`.github/workflows/build-fleetctl-pkg.yml`) for manually building and testing the macOS `.pkg` installer. This workflow allows team members to trigger test builds on demand, supports both test and production modes, and ensures the package is properly signed and notarized before being uploaded as an artifact or attached to a release.

# Checklist for submitter

If some of the following don't apply, delete the relevant line.

- [x] Changes file added for user-visible changes in `changes/`, `orbit/changes/` or `ee/fleetd-chrome/changes`.
  See [Changes files](https://github.com/fleetdm/fleet/blob/main/docs/Contributing/guides/committing-changes.md#changes-files) for more information.

## Testing

- [ ] Added/updated automated tests
- [X] QA'd all new/changed functionality manually
  - [X] Tested using the included build-fleetctl-pkg.yml, downloaded and installed the package on a Mac VM and it worked!

It's very difficult to e2e test the goreleaser workflow without doing an actual release (we could do a prerelease and then delete it, but it's not ideal).  But the benefit of having the separate packager workflow is that we can run it at any time with test mode turned off to push a package to a release.  So if the gorelease workflow doesn't work right out of the gate, we can use the separate workflow to get the same effect.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * fleetctl is now distributed as a signed, notarized macOS .pkg installer for easier installation.
  * CI/CD can build the macOS package automatically and (in test mode) upload artifacts for review.
* **Documentation**
  * Added note describing the new macOS .pkg distribution format.
* **Chores**
  * Release tooling updated to conditionally trigger package builds when signed macOS binaries are available.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

allenhouchins committed:
Create build-fleetctl-pkg.yml

allenhouchins committed:
Add test mode to fleetctl macOS package workflow

Introduces a 'test_mode' input for manual workflow runs, allowing package builds to be tested without uploading to releases. In test mode, the package is uploaded as an artifact instead. Also updates version extraction logic to support test builds triggered from branches.

allenhouchins committed:
Update build-fleetctl-pkg.yml

allenhouchins committed:
Update build-fleetctl-pkg.yml

allenhouchins committed:
Update build-fleetctl-pkg.yml

allenhouchins committed:
Update build-fleetctl-pkg.yml

allenhouchins committed:
Update build-fleetctl-pkg.yml

allenhouchins committed:
Rename macOS package output to include '_mac' suffix

Updated workflow steps to use 'fleetctl_v<version>_mac.pkg' instead of 'fleetctl_v<version>.pkg' for macOS package outputs. This change ensures clearer package identification and avoids naming conflicts with other platform builds.

allenhouchins committed:
integrate with goreleaser

allenhouchins committed:
Changes to make ready for production

allenhouchins committed:
Create 35089-fleetctl-released-as-pkg

lukeheath commented:
@allenhouchins This is really cool, thanks for adding it! I'll defer to @lucasmrod for review but let me know if my approval is needed. 

One note: we can't effectively test changes to the goreleaser flows without actually issuing a release, so it's possible that if there's an issue it will be release blocking and need rapid troubleshooting.

allenhouchins committed:
Improve notarization flow and release uploads

Refactor notarization parsing and retry logic, fix binary invocation, and make GH release uploads idempotent.

Changes:
- .github/workflows/build-fleetctl-pkg.yml: removed fallback echoes for make steps, adjusted status parsing (use last status line), reworked notarization success/failure branches to fetch logs on non-Accepted statuses, preserve test-mode behavior, and add --clobber to gh release upload with retry.
- .goreleaser.yml: use POSIX-style uname check ([ ... ] and =) for macOS conditional before running the pkg builder.
- tools/build-fleetctl-pkg/main.sh: invoke the resolved signed binary path for --version, use tail to pick the last notarization status, simplify loop to break on Accepted and otherwise fetch logs and retry until attempts exhausted, and use --clobber for gh release uploads.

These changes make notarization detection and failure handling more robust, ensure the correct binary is used, and make release uploads reliable on retry.

allenhouchins committed:
Quote secrets, gate pkg build, and default signer

Quote keychain and certificate password variables in the mac package workflow to avoid issues with spaces and ensure secure commands. Make PACKAGE_SIGNING_IDENTITY_SHA1 overridable via an environment variable (with the existing SHA1 as the default) in both the workflow and the packaging script. Add a goreleaser job output (binary_uploaded) and conditionally run the mac build-fleetctl-pkg job only when the signed binary exists; also ensure outputs are written to GITHUB_OUTPUT correctly.

allenhouchins committed:
Apply suggestions from code review

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

sgress454 committed:
fix syntax error

sgress454 committed:
consolidate test workflow and script

sgress454 committed:
update quoting

sgress454 committed:
cleanup

sgress454 committed:
setup node

sgress454 committed:
extract identities rather than hard-coding

sgress454 committed:
fix path

sgress454 committed:
remove unnecessary gh install

sgress454 committed:
remove post-hook in favor of the job in the goreleaser workflow

sgress454 committed:
address review comments

sgress454 commented on .github/workflows/build-fleetctl-pkg.yml:
```diff
@@ -0,0 +1,226 @@
+          REF_NAME="${{ github.ref_name }}"
+
+          # Check if running from a tag (for testing tag-based workflows)
+          if [[ "${{ github.ref }}" == refs/tags/* ]] && [[ "$REF_NAME" == fleet-* ]]; then
+            # Extract version from tag (e.g., fleet-v4.10.0 -> v4.10.0)
+            VERSION="${REF_NAME#fleet-}"
+            TAG_NAME="$REF_NAME"
+          else
+            # For branch-based manual triggers, use a test version
+            VERSION="test-$(date +%Y%m%d-%H%M%S)"
+            TAG_NAME="fleet-${VERSION}"
```
good catch! fixed

sgress454 commented on .github/workflows/build-fleetctl-pkg.yml:
```diff
@@ -0,0 +1,226 @@
+          GITHUB_REPOSITORY: ${{ github.repository }}
+          GITHUB_REF: ${{ github.ref }}
+          VERSION: ${{ steps.extract_version.outputs.version }}
+          SKIP_UPLOAD: ${{ steps.extract_version.outputs.is_test_mode }}
```
added validation for tag 👍

sgress454 committed:
fail if attestation fails on pkg workflow

sgress454 commented on .github/workflows/build-fleetctl-pkg.yml:
```diff
@@ -0,0 +1,226 @@
+          retention-days: 7
+
+      - name: Attest package
+        continue-on-error: true
+        uses: actions/attest-build-provenance@619dbb2e03e0189af0c55118e7d3c5e129e99726 # v2.0
```
good call -- since this is the last step, i'm fine getting the fail signal here.

sgress454 commented on tools/build-fleetctl-pkg/main.sh:
```diff
@@ -0,0 +1,223 @@
+echo "Found signed binary at: $FLEETCTL_BINARY"
+"$FLEETCTL_BINARY" --version
+
+cleanup() {
+    echo "Cleaning up..."
+    rm -f installer_certificate.p12
+    rm -f fleetctl_unsigned.pkg
+    rm -f fleetctl_notarize.zip
+    rm -rf pkgroot
+}
+
+# Trap EXIT signal to call cleanup function
+trap cleanup EXIT
+
+# Import package signing keys (for .pkg)
+echo "Importing package signing certificate..."
+echo "$APPLE_INSTALLER_CERTIFICATE" | base64 --decode > installer_certificate.p12
+security create-keychain -p "$KEYCHAIN_PASSWORD" build.keychain || true
+security default-keychain -s build.keychain
+security unlock-keychain -p "$KEYCHAIN_PASSWORD" build.keychain
+security import installer_certificate.p12 -k build.keychain -P "$APPLE_INSTALLER_CERTIFICATE_PASSWORD" -T /usr/bin/productsign
+security set-key-partition-list -S apple-tool:,apple:,productsign: -s -k "$KEYCHAIN_PASSWORD" build.keychain
+security find-identity -vv
```
this is run by ephemeral runners, no need for extra cleanup

sgress454 commented on tools/build-fleetctl-pkg/main.sh:
```diff
@@ -0,0 +1,223 @@
+
+# Move package to dist directory so it can be picked up by goreleaser or uploaded later
+mkdir -p dist
+mv "$PACKAGE_NAME" "dist/$PACKAGE_NAME"
+
+echo "✓ Package created successfully: dist/$PACKAGE_NAME"
+
+# Upload to release unless skipped
+if [[ "$SKIP_UPLOAD" == "true" ]]; then
+    echo "Skipping release upload (SKIP_UPLOAD=true)"
+elif [[ -n "$GITHUB_TOKEN" ]] && [[ -n "$GITHUB_REPOSITORY" ]] && command -v gh &> /dev/null; then
+    TAG_NAME="${GITHUB_REF#refs/tags/}"
+    echo "Uploading package to release $TAG_NAME..."
```
fixed

sgress454 committed:
get version from script arg, not env var

iansltx commented on .github/workflows/build-fleetctl-pkg.yml:
```diff
@@ -0,0 +1,226 @@
+# 1. Go to Actions → Build fleetctl macOS package → Run workflow
+# 2. Select your branch (e.g., main or feature branch)
+# 3. Test mode defaults to true (recommended) - package will be built and uploaded as an artifact
+# 4. To test release upload, set test_mode to false (use with caution)
```
Probably want to lock this down access-wise or something, as IIRC the goreleaser action is.

sgress454 replied:
goreleaser action doesn't have a manual dispatch option at all.  But locking this and the Windows version down probably makes sense, I'll ping @lukeheath on it.

iansltx commented on .github/workflows/build-fleetctl-pkg.yml:
```diff
@@ -0,0 +1,224 @@
+          echo "$APPLE_APPLICATION_CERTIFICATE" | base64 --decode > app_certificate.p12
+          security create-keychain -p "$KEYCHAIN_PASSWORD" build.keychain || true
+          security default-keychain -s build.keychain
+          security unlock-keychain -p "$KEYCHAIN_PASSWORD" build.keychain
```
Assuming GitHub elides these in logs?

sgress454 replied:
[jes](https://github.com/fleetdm/fleet/actions/runs/24246732981/job/70794924810#step:11:20)

<img width="611" height="117" alt="Image" src="https://github.com/user-attachments/assets/323ce737-8402-4653-a7ba-d57d33c722b2" />

sgress454 replied:
(the security commands themselves have no output)

iansltx commented on .github/workflows/build-fleetctl-pkg.yml:
```diff
@@ -0,0 +1,224 @@
+          chmod +x tools/build-fleetctl-pkg/main.sh
+          ./tools/build-fleetctl-pkg/main.sh ./fleetctl "${{ steps.extract_version.outputs.version }}"
+
+      - name: Upload package artifact (test mode only)
```
So here we're uploading this as a build artifact, rather than pushing it to the release? Feels a bit weird to split the behaviors such that one upload is in an action step and the other is in a script, and that may have been cause for my confusion when looking at the Windows PR just now. Or we're handling the Windows upload differently than this one?

sgress454 replied:
We're pushing the artifact only in test mode (so when we _don't_ push it to the release).  I updated the Windows PR to match this one, where the `SKIP_UPLOAD` env var is set based on the test mode value, and `main.sh` uses that to determine whether to push to release or not.

iansltx commented on tools/build-fleetctl-pkg/main.sh:
```diff
@@ -0,0 +1,228 @@
+    
+    if [ $ELAPSED -ge $MAX_WAIT ]; then
+        echo "Warning: Release not found after waiting ${MAX_WAIT}s"
+        echo "Attempting to upload anyway (release might be created as draft)..."
```
Does this mean that in a goreleaser workflow this will wait the full 5 mins and then push?

sgress454 replied:
It'll wait _up_ to five minutes and then try to push.  If it finds a release on line 202 then ELAPSED won't get incremented so the ` if [ $ELAPSED -ge $MAX_WAIT ];` will fail, and we won't see this log.

sgress454 replied:
@lucasmrod I think the thinking here was that there could be lag between when goreleaser publishes the release and when `gh release view` shows it, but maybe that's not a concern? 

Do we ever publish draft releases in another way, where we'd end up waiting this whole 5 minutes before pushing the package?

lukeheath commented:
@copilot resolve the merge conflicts in this pull request

Copilot committed:
Merge origin/main and resolve conflicts in goreleaser-fleet.yaml

Co-authored-by: lukeheath <2495927+lukeheath@users.noreply.github.com>

lukeheath commented:
@coderabbitai full review

lukeheath commented on .github/workflows/goreleaser-fleet.yaml:
```diff
@@ -193,6 +212,65 @@ jobs:
+
+      - name: Checkout
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3.5.3
+        with:
+          fetch-depth: 0
+
+      - name: Download signed fleetctl binary
+        uses: actions/download-artifact@5e317d1137c093c30a278992a91e9db4e3c1f039 # v4.1.8
+        with:
+          name: fleetctl-signed-macos
+          path: dist/fleetctl_darwin_all
+
+      - name: Verify binary exists
+        run: |
+          if [[ ! -f "dist/fleetctl_darwin_all/fleetctl" ]]; then
+            echo "Error: Signed fleetctl binary not found"
+            exit 1
+          fi
+          echo "Binary found, verifying signature..."
+          codesign --verify --verbose dist/fleetctl_darwin_all/fleetctl || exit 1
+          ./dist/fleetctl_darwin_all/fleetctl --version
+
```
Valid. Fixed.

sgress454 replied:
good catch, fixed

lukeheath commented on .github/workflows/build-fleetctl-pkg.yml:
```diff
@@ -0,0 +1,224 @@
+# handles building and uploading the fleetctl package automatically when a fleet-* tag is pushed.
+# This workflow is kept for manual testing only.
+#
+# TESTING:
+# To test the package build process:
+# 1. Go to Actions → Build fleetctl macOS package → Run workflow
+# 2. Select your branch (e.g., main or feature branch)
+# 3. Test mode defaults to true (recommended) - package will be built and uploaded as an artifact
+# 4. To test release upload, set test_mode to false (use with caution)
```
Valid. Fixed.

sgress454 replied:
@coderabbitai The intended use case for `test_mode=false` is running this workflow on an already-tagged branch that for some reason failed the main goreleaser workflow.  It will work for that use case, and fail (with an acceptable error) for others.  We can update if we ever want to allow pushing artifacts from non-tagged branches to a release, but more changes would need to be made in that case.

lukeheath commented:
@sgress454 Looks good to me. The CodeRabbit comments seem worth looking at so holding off on approval. Just let me know when it's ready.

sgress454 committed:
add chmod

sgress454 committed:
comment
